core/src/main/java/tc/oc/pgm/trigger/TriggerParser.java

```diff
@@ -0,0 +1,165 @@
+import tc.oc.pgm.trigger.triggers.ChatMessageTrigger;
+import tc.oc.pgm.trigger.triggers.KitTrigger;
+import tc.oc.pgm.trigger.triggers.TriggerNode;
+import tc.oc.pgm.util.MethodParser;
+import tc.oc.pgm.util.MethodParsers;
+import tc.oc.pgm.util.xml.InvalidXMLException;
+import tc.oc.pgm.util.xml.Node;
+import tc.oc.pgm.util.xml.XMLUtils;
+
+public class TriggerParser {
+
+  private final MapFactory factory;
+  private final FilterParser filters;
+  private final Map<String, Method> methodParsers;
+
+  public TriggerParser(MapFactory factory) {
+    this.factory = factory;
+    this.filters = factory.getFilters();
+    this.methodParsers = MethodParsers.getMethodParsersForClass(getClass());
+  }
+
+  public <B extends Filterable<?>> Trigger<? super B> parse(Element el, @Nullable Class<B> bound)
+      throws InvalidXMLException {
+    String id = FeatureDefinitionContext.parseId(el);
+
+    if (id != null && maybeReference(el)) {
+      return parseReference(new Node(el), id, bound);
+    }
+
+    TriggerDefinition<? super B> result =
+        "trigger".equals(el.getName()) && bound != null
+            ? parseDefinitionBounded(el, bound)
+            : parseDynamic(el);
+
+    Class<?> childBound = result.getTriggerType();
+    if (bound != null && !childBound.isAssignableFrom(bound)) {
+      throw new InvalidXMLException(
+          "Wrong trigger target, expected "
+              + bound.getSimpleName()
+              + " rather than "
+              + childBound.getSimpleName(),
+          new Node(el));
+    }
+
+    factory.getFeatures().addFeature(el, result);
+    return result;
+  }
+
+  public boolean isTrigger(Element el) {
+    return getParserFor(el) != null;
+  }
+
+  private boolean maybeReference(Element el) {
+    return "trigger".equals(el.getName()) && el.getChildren().isEmpty();
+  }
+
+  public <B> Trigger<? super B> parseReference(Node node, Class<B> bound)
+      throws InvalidXMLException {
+    return parseReference(node, node.getValue(), bound);
+  }
+
+  public <B> Trigger<? super B> parseReference(Node node, String id, Class<B> bound)
+      throws InvalidXMLException {
+    return factory
+        .getFeatures()
+        .addReference(new XMLTriggerReference<>(factory.getFeatures(), node, id, bound));
+  }
+
+  private <B extends Filterable<?>> TriggerDefinition<? super B> parseDefinitionBounded(
+      Element el, @Nonnull Class<B> bound) throws InvalidXMLException {
+    ImmutableList.Builder<Trigger<? super B>> builder = ImmutableList.builder();
+    for (Element child : el.getChildren()) {
+      builder.add(parse(child, bound));
+    }
+
+    return new TriggerNode<>(builder.build(), bound);
+  }
+
+  protected Method getParserFor(Element el) {
+    return methodParsers.get(el.getName().toLowerCase());
+  }
+
+  private <T> TriggerDefinition<T> parseDynamic(Element el) throws InvalidXMLException {
+    Method parser = getParserFor(el);
+    if (parser != null) {
+      try {
+        //noinspection unchecked
```

Comment:
Maybe just a `@SuppressWarnings("unchecked")` instead?

Reply:
that needs to be on a method, class, or field level, a return X can't have that, so i'd need to make it for the whole method

Reply:
Yeah my thought was the `praseDynamic` method